### ISSUE TITLE
Error in mdp_apps.ipynb

### ISSUE THREAD

**Issue by aakashdeep791** [aimacode/aima-python#951]
While running 

mdp = CustomMDP(t, rewards, terminals, init, gamma=.9)

getting this error:- 

['sleep', 'facebook', 'quit', 'study', 'pub']
Could not retrieve states from transitions
Warning: Transition table is empty.
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-7-09d6e6e70a9c> in <module>()
----> 1 mdp = CustomMDP(t, rewards, terminals, init, gamma=.9)

<ipython-input-6-2a2d20bab823> in __init__(self, transition_matrix, rewards, terminals, init, gamma)
      9         print(actlist)
     10 
---> 11         MDP.__init__(self, init, actlist, terminals=terminals, gamma=gamma)
     12         self.t = transition_matrix
     13         self.reward = rewards

D:\Australia\UOW\Spring 2018\Reasoning and Learning\assignment_2\aima-python-master\aima-python-master\mdp.py in __init__(self, init, actlist, terminals, transitions, reward, states, gamma)
     48         self.gamma = gamma
     49 
---> 50         self.reward = reward or {s: 0 for s in self.states}
     51 
     52         # self.check_consistency()

TypeError: 'NoneType' object is not iterable

**Comment by ad71** [comment]
In `CustomMDP` there is a line that calls the constructor of the base class. Replace it with:
```python
MDP.__init__(self, init, actlist, terminals, transition_matrix, rewards, gamma=gamma)
```
Tell me if this doesn't work.

**Comment by aakashdeep791** [comment]
Yes, its working now, thanks. But, there are still some errors in the later code.

In valueiteration(mdp)    getting this error:-

---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-8-0a7baeac31c2> in <module>()
----> 1 value_iteration(mdp)

D:\Australia\UOW\Spring 2018\Reasoning and Learning\assignment_2\aima-python-master\aima-python-master\mdp.py in value_iteration(mdp, epsilon)
    208         delta = 0
    209         for s in mdp.states:
--> 210             U1[s] = R(s) + gamma * max(sum(p*U[s1] for (p, s1) in T(s, a))
    211                                                    for a in mdp.actions(s))
    212             delta = max(delta, abs(U1[s] - U[s]))

D:\Australia\UOW\Spring 2018\Reasoning and Learning\assignment_2\aima-python-master\aima-python-master\mdp.py in R(self, state)
     55         """Return a numeric reward for this state."""
     56 
---> 57         return self.reward[state]
     58 
     59     def T(self, state, action):

KeyError: 'l'

same kind of errror in:-      pi = best_policy(mdp, value_iteration(mdp, .01))

**Comment by khatiwadaaj** [comment]
aakashdeep91: have you got any further with this?

**Comment by aakashdeep791** [comment]
I found a workaround, but not sure whether its correct.

In mdp.py file, if you update the "get_states_from_transitions" method with the below code, it should give the expected output.

 def get_states_from_transitions(self, transitions):
        if isinstance(transitions, dict):
            s1 = set(transitions.keys())
            return s1
        else:
            print('Could not retrieve states from transitions')
            return None

**Comment by dmeoli** [comment]
Fixed in `ca11f95`. The root cause was `CustomMDP.__init__` calling `MDP.__init__` without `states`/`transitions`, so `get_states_from_transitions(None)` left `self.states = None` and the reward default `{s: 0 for s in self.states}` raised `TypeError: 'NoneType' object is not iterable` — which then cascaded into the `NameError: name 'mdp' is not defined` failures in the later cells. The notebook also still imported the removed `utils.argmax` (now `argmax_random_tie`).

Both are fixed and the notebook now runs top-to-bottom with 0 errors. Thanks for the report!
